miden-lib/asm/sat/prologue.masm

```diff
@@ -0,0 +1,416 @@
+    padw padw padw
+
+    # read the account data
+    adv_pipe adv_pipe
+
+    # extract digest from hasher rate elements (h_0, ..., h_3)
+    dropw swapw dropw
+
+    # assert that the account hash matches the hash in global inputs
+    padw push.ACCT_HASH_PTR mem_loadw assert_eqw
+
+    # assert the account id matches the account id in global inputs
+    push.ACCT_ID_PTR mem_load
+    push.ACCT_ID_AND_NONCE_PTR mem_load
+    assert_eq
+
+    # clear the stack
+    drop
+end
+
+# CONSUMED NOTES DATA
+# =================================================================================================
+
+#! Computes a pointer to the memory address at which the data associated with a note with index i
+#! is stored.
+#!
+#! Stack: [i]
+#! Output: [ptr]
+#!
+#! - i is the index of the consumed note.
+#! - ptr is the memory address of the data segment for note i.
+proc.consumed_note_data_ptr
+    push.1 add push.CONSUMED_NOTE_MEM_SIZE mul push.CONSUMED_NOTE_SECTION_OFFSET add
+end
+
+#! Computes a pointer to the memory address at which the nullifier associated a note with index i
+#! is stored.
+#!
+#! Stack: [i]
+#! Output: [ptr]
+#!
+#! - i is the index of the consumed note.
+#! - ptr is the nullifier memory address for note i.
+proc.consumed_note_nullifier_ptr
+    push.CONSUMED_NOTE_SECTION_OFFSET.1 add add
+end
+
+#! Reads data for consumed note i from the advice provider and stores it in memory at the
+#! appropriate memory address. This includes computing and storing the nullifier and the
+#! note hash.
+#!
+#! Stack: [i]
+#! Advice Tape: [CN1_SN, CN1_SR, CN1_IR, CN1_VR,
+#!               CN1_NA,
+#!               CN1_A1, CN1_A2, ...]
+#!
+#! Output: []
+#!
+#! - i is the index of the consumed note.
+#! - CN1_SN is the serial number of consumed note 1.
+#! - CN1_SR is the script root of consumed note 1.
+#! - CN1_IR is the inputs root of consumed note 1.
+#! - CN1_VR is the vault root of consumed note 1.
+#! - CN1_NA is the number of assets in consumed note 1.
+#! - CN1_A1 is the first asset of consumed note 1.
+#! - CN1_A2 is the second asset of consumed note 1.
+proc.process_consumed_note
+    # read core note data
+    # ---------------------------------------------------------------------------------------------
+
+    # compute address to store note hash
+    exec.consumed_note_data_ptr
+
+    # compute address to store raw note data
+    dup push.1 add
+
+    # read note data from the advice provider
+    padw padw padw
+    adv_pipe adv_pipe
+
+    # extract nullifier
+    dropw swapw dropw
+
+    # compute address for nullifier
+    # dup.6 = counter / note index
+    dup.6 exec.consumed_note_nullifier_ptr
+
+    # store nullifier in memory and drop from stack
+    mem_storew dropw
+
+    # ingest note assets
+    # ---------------------------------------------------------------------------------------------
+
+    # read the number of assets from the advice provider and store in memory
+    adv_push.1 dup dup.2 mem_store
```

Comment:
Should we explicitly check if the number of assets is smaller than w/e limit we impose?

Reply:
assertion added with limit of 256.